### PR TITLE
exp/memory: add multiple fixes

### PR DESCRIPTION
**Pull request by FluffyKebab**
-  Make buffer memory implement schema.Memory
 - Add comments for all exported symbols
 - Statically assert that the memory implementations implements the memory interface
 - Update tests to use t.Fatal
 - Change schema.Memory LoadMemoryVariables method to return error because the buffer memory needs to return an error
 - Rename from BufferMemory and EmptyMemory to Buffer and Empty to remove repetition

### PR Checklist

- [x] Read the [Contributing documentation](https://github.com/tmc/langchaingo/blob/main/CONTRIBUTING.md).
- [x] Read the [Code of conduct documentation](https://github.com/tmc/langchaingo/blob/main/CODE_OF_CONDUCT.md).
- [x] Name your Pull Request title clearly, concisely, and prefixed with the name of primarily affected package you changed according to [Good commit messages](https://go.dev/doc/contribute#commit_messages) (such as `memory: add interfaces for X, Y` or `util: add whizzbang helpers`).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [x] Describes source of new concepts.
- [x] References existing implementations as appropriate.
- [x] Contains test coverage for new functions.
- [x] Passes all [`golangci-lint`](https://golangci-lint.run/) checks.

### CODE REVIEW

File: exp/chains/chains.go [hunk 1]
```diff
@@ -2,19 +2,23 @@ package chains
...
 	fullValues := make(map[string]any, 0)
-
 	for key, value := range inputValues {
 		fullValues[key] = value
 	}
 
-	newValues := c.GetMemory().LoadMemoryVariables(inputValues)
+	newValues, err := c.GetMemory().LoadMemoryVariables(inputValues)
+	if err != nil {
+		return map[string]any{}, err
```
Comment (tmc) [inline exp/chains/chains.go:20]: typically you'd return nil here

File: exp/chains/llm_chain.go [hunk 2]
```diff
@@ -22,7 +22,7 @@ func NewLLMChain(llm llms.LLM, prompt prompts.Template) LLMChain {
 		llm:          llm,
 		OutputKey:    "text",
 		OutputParser: output_parsers.NewEmptyOutputParser(),
-		Memory:       memory.NewEmptyMemory(),
+		Memory:       memory.NewEmpty(),
```
Comment (tmc) [inline exp/chains/llm_chain.go:25]: good

File: exp/memory/buffer_memory.go [hunk 3]
```diff
@@ -1,8 +1,15 @@
...
-// BufferMemory is a simple form of memory that remembers previous conversational back and forths directly.
-type BufferMemory struct {
-	ChatHistory    *ChatMessageHistory
+import (
+	"fmt"
+
+	"github.com/tmc/langchaingo/schema"
+)
+
+// memory.Buffer is a simple form of memory that remembers previous conversational back and forths directly.
```
Comment (tmc) [inline exp/memory/buffer_memory.go:9]: this should just be "Buffer" not "memory.Buffer"

File: exp/memory/buffer_memory.go [hunk 4]
```diff
@@ -11,7 +18,54 @@ type BufferMemory struct {
 	MemoryKey      string
 }
 
-func (m BufferMemory) SaveContext(inputValues map[string]any, outputValues map[string]any) error {
+// statically assert that BufferMemory implement the memory interface:
+var _ schema.Memory = &Buffer{}
+
+// Creates new buffer memory
```
Comment (tmc) [inline exp/memory/buffer_memory.go:24]: please end sentences with periods

**Review by tmc — APPROVED** [review]
Good changes, I have input on the comment styles that I'd like you to address but we can do that in follow-up.

**Comment by FluffyKebab** [comment]
@tmc Addressed your initial concerns. How should I improve the comments?

**Comment by tmc** [comment]
@FluffyKebab can you rebase this?

**Pull request closed without merging**